e2e/testsuite/testconfig.go

```diff
@@ -207,6 +216,35 @@ func (tc TestConfig) validateGenesisDebugConfig() error {
 	return err
 }
 
+// validateUpgradeConfig ensures the upgrade configuration is valid.
+func (tc TestConfig) validateUpgradeConfig() error {
+	if strings.TrimSpace(tc.UpgradePlanName) == "" {
+		return nil
+	}
+
+	// the upgrade plan name specified must match one of the upgrade plans in the upgrade configs.
+	foundPlan := false
+	for _, upgrade := range tc.UpgradeConfigs {
+		if strings.TrimSpace(upgrade.Tag) == "" {
+			return fmt.Errorf("upgrade config missing tag: %+v", upgrade)
+		}
+
+		if strings.TrimSpace(upgrade.PlanName) == "" {
+			return fmt.Errorf("upgrade config missing plan name: %+v", upgrade)
+		}
```

Comment:
[nit] I wonder if it would make more sense to either:

1) do this check only if `upgrade.PlanName == tc.UpgradePlanName` (so a "broken" upgrade config is allowed if it's not used for the test)
2) split this into two different loops. Since we don't care about efficiency here, I think it's a bit more clear if validation and search are split

Reply:
I figured it would be best to just explode immediately if anything about config is bad. Don't have a strong preference for multiple loops, will always just be a few elements anyway.

Reply:
Up to you! I am slightly more inclined towards having one loop of validation followed by one of search, for clarity, but not a deal breaker